Neos.Neos/Classes/Fusion/Cache/GraphProjectorCatchUpHookForCacheFlushing.php

```diff
@@ -173,66 +222,104 @@ public function onAfterEvent(EventInterface $eventInstance, EventEnvelope $event
+            $nodeAggregate->nodeAggregateId,
+            $nodeAggregate->nodeTypeName,
+            $this->determineParentNodeAggregateIds($workspaceName, $nodeAggregate->nodeAggregateId, NodeAggregateIds::createEmpty())
+        );
     }
 
-    public function onBeforeBatchCompleted(): void
+    private function scheduleCacheFlushJobForWorkspaceName(
+        ContentRepository $contentRepository,
+        WorkspaceName $workspaceName
+    ): void {
+        // we store this in an associative array deduplicate.
+        $this->flushWorkspaceRequestsOnBeforeBatchCompleted[$workspaceName->value] = FlushWorkspaceRequest::create(
+            $contentRepository->id,
+            $workspaceName,
+        );
+    }
+
+    private function determineParentNodeAggregateIds(WorkspaceName $workspaceName, NodeAggregateId $childNodeAggregateId, NodeAggregateIds $collectedParentNodeAggregateIds): NodeAggregateIds
     {
-        foreach ($this->cacheFlushesOnBeforeBatchCompleted as $k => $entry) {
-            $this->contentCacheFlusher->flushNodeAggregate($entry['cr'], $entry['wsn'], $entry['nai']);
-            $this->cacheFlushesOnAfterCatchUp[$k] = $entry;
+        $parentNodeAggregates = $this->contentRepository->getContentGraph($workspaceName)->findParentNodeAggregates($childNodeAggregateId);
+        $parentNodeAggregateIds = NodeAggregateIds::fromArray(
+            array_map(static fn (NodeAggregate $parentNodeAggregate) => $parentNodeAggregate->nodeAggregateId, iterator_to_array($parentNodeAggregates))
+        );
+
+        foreach ($parentNodeAggregateIds as $parentNodeAggregateId) {
+            // Prevent infinite loops
+            if (!$collectedParentNodeAggregateIds->contain($parentNodeAggregateId)) {
+                $collectedParentNodeAggregateIds = $collectedParentNodeAggregateIds->merge(NodeAggregateIds::create($parentNodeAggregateId));
+                $collectedParentNodeAggregateIds = $this->determineParentNodeAggregateIds($workspaceName, $parentNodeAggregateId, $collectedParentNodeAggregateIds);
+            }
         }
-        $this->cacheFlushesOnBeforeBatchCompleted = [];
+
+        return $collectedParentNodeAggregateIds;
     }
 
+    public function onBeforeBatchCompleted(): void
+    {
+        foreach ($this->flushNodeAggregateRequestsOnBeforeBatchCompleted as $index => $request) {
+            $this->contentCacheFlusher->flushNodeAggregate($request, CacheFlushingStrategy::IMMEDIATELY);
+            $this->flushNodeAggregateRequestsOnAfterCatchUp[$index] = $request;
+        }
+        $this->flushNodeAggregateRequestsOnBeforeBatchCompleted = [];
+
+        foreach ($this->flushWorkspaceRequestsOnBeforeBatchCompleted as $index => $request) {
+            $this->contentCacheFlusher->flushWorkspace($request, CacheFlushingStrategy::IMMEDIATELY);
+            $this->flushWorkspaceRequestsOnAfterCatchUp[$index] = $request;
+        }
+        $this->flushWorkspaceRequestsOnBeforeBatchCompleted = [];
+    }
```

Comment:
We might need to discuss this "double-cache-flush" approach. See Sebastians description in the head of this file (eventual consistency) vs. performance considerations.